test/unit/Cardano/Wallet/CoinSelection/LargestFirstSpec.hs

```diff
@@ -0,0 +1,336 @@
+    result <- runExceptT $ largestFirst
+              (defaultCoinSelectionOptions 100)
+              utxo
+              txOuts
+
+    L.length
+        (concatMap (\ (CoinSelection inps _ _) -> inps) $ rights [result])
+        `shouldSatisfy`
+        (>= NE.length txOuts)
+
+propInputDecreasingOrder
+    :: CoveringCase
+    -> Property
+propInputDecreasingOrder (CoveringCase (utxo, txOuts)) =
+    (condCoinsCovering txOuts utxo) ==> monadicIO $ liftIO $ do
+    result <- runExceptT $ largestFirst
+              (defaultCoinSelectionOptions 100)
+              utxo
+              txOuts
+    let utxoEntriesInCoinSelection =
+            concatMap (\ (CoinSelection inps _ _) -> inps)
+            $ rights [result]
+    let utxo' =
+            Map.toList (getUTxO utxo) L.\\ utxoEntriesInCoinSelection
+    let getExtremumValue f = f . map (getCoin . coin . snd)
+
+    if L.null utxo' then utxo' `shouldBe` utxo' else
+        (getExtremumValue L.minimum utxoEntriesInCoinSelection)
+        `shouldSatisfy` (>= (getExtremumValue L.maximum utxo'))
+
+{-------------------------------------------------------------------------------
+                                  Test Data
+-------------------------------------------------------------------------------}
+
+defaultCoinSelectionOptions
+    :: Word64
+    -> CoinSelectionOptions
+defaultCoinSelectionOptions n = CoinSelectionOptions
+    { estimateFee = \_ _ -> Coin 0
+    , dustThreshold = Coin 0
+    , maximumNumberOfInputs = n
+    }
+
+
+-- Check if there are enough covering utxo entries
+-- in the utxo to cover all transaction outputs
+condCoinsCovering
+    :: NonEmpty TxOut
+    -> UTxO
+    -> Bool
+condCoinsCovering txOuts utxo =
+    let
+        len = NE.length txOuts
+    in L.length (utxoCovering txOuts utxo) >= len
+
+-- Select all UTxO entries that could cover all
+-- transaction outputs sorted from the smallest one
+utxoCovering
+    :: NonEmpty TxOut
+    -> UTxO
+    -> [(TxIn, TxOut)]
+utxoCovering txOuts utxo =
+    L.sortOn (coin . snd)
+    $ L.foldl (\acc o ->
+                   rotate 1
+                   $  L.dropWhile (\(_,_o) -> coin _o < coin o) acc
+              ) utxoSorted
+    $ outputsSorted txOuts
+    where
+        utxoSorted = L.sortOn (coin . snd)
+                     $ (Map.toList . getUTxO) utxo
+        rotate :: Int -> [a] -> [a]
+        rotate _ [] = []
+        rotate n xs = zipWith const (drop n (cycle xs)) xs
+
+-- Sort transaction outputs in the increasing order
+-- as we want to cover smaller payments first
+outputsSorted
+    :: NonEmpty TxOut
+    -> [TxOut]
+outputsSorted txOuts =
+    L.sortOn coin $ NE.toList txOuts
+
+newtype CoveringCase = CoveringCase { getCoveringCase :: (UTxO, NonEmpty TxOut)}
+    deriving Show
+
+instance Arbitrary CoveringCase where
+    arbitrary = do
+        n <- choose (1, 2)
+        txOutsNonEmpty <- NE.fromList <$> vectorOf n arbitrary
+        utxo <- arbitrary `suchThat` (condCoinsCovering txOutsNonEmpty)
```

Comment:
Hmmm... we should have this either as a precondition or in the generator. Having it in both places is redundant. I'd be in favor of generating only valid `CoveringCase`, so checking only here once (previous comment about simplifying the precondition still applies though).

Reply:
I like it to have in generator, but when I would know how to put `isRight` in the property precondition I would remove `suchThat`